### ISSUE TITLE
[css-images-4] Clarifying CSS gradient rendering for edge cases involving "longer hue" interpolation

### ISSUE THREAD

**Issue by jfkthame** [w3c/csswg-drafts#11381]
I am seeing differences between Firefox, Chrome and Safari regarding how gradients with the "longer hue" interpolation method are rendered. I think I understand what the spec expects, but none of the browsers tested get this right, so I'd like to confirm my interpretation — and perhaps some clarification or additional examples may be needed in the spec.

Testcase: https://codepen.io/jfkthame/pen/azopZOx

(Note that there have been some recent patches to gradient rendering in Firefox. The description below refers to current Nightly builds; older versions have some different and less self-consistent results.)

This renders a number of 100px-wide gradient swatches, all of them defined as linear-gradient(to right in hsl longer hue, ...) with two color stops, red and blue, at a variety of positions. (The red and blue tick-marks above and below each swatch indicate the positions of the color stops defining the gradient.)

Results I'm seeing in Firefox Nightly, Chrome Canary, and Safari Tech Preview:
<img width="746" alt="image" src="https://github.com/user-attachments/assets/7c7661b1-9185-418c-8475-6def180120e4" />

The simplest case is (a), with the red stop at the left edge and the blue stop at the right. No issues here: we simply render a gradient from red to blue, going "the long way around" the HSL wheel.

What happens if the first and last stops defining the gradient do not cover the full extent of the area to be rendered? In (b), the blue stop is in the middle, so what happens in the right-hand half of the swatch? According to https://drafts.csswg.org/css-images-4/#coloring-gradient-line,

> Before the first color stop, the gradient line is the color of the first color stop, and after the last color stop, the gradient line is the color of the last color stop.

which seems to imply that the right-hand half of (b) should be solid blue. However, all three tested browsers actually paint a further gradient here, with a full cycle of the HSL wheel from the blue stop at 50px all the way around to blue again at the 100px edge. I presume the browsers are adding an "implicit" blue color stop at the right-hand edge, and then performing a longer-hue interpolation between the actual blue stop at 50px and that implicit one. I don't see any justification in the spec for this behavior: I think it's a bug, but given that all three browsers behave this way, maybe I'm missing something?

Example (c) is similar: here, both the red and blue color stops are moved in from the edges of the swatch, so I think the expected rendering should have 20px of solid red at the left, and 25px of blue at the right, with the longer-hue gradient from red to blue in between. But all three browsers also render a complete 360° gradient in both the left and right portions, outside the defined color stops. Bug?

(Aside: the description of linear-gradient on MDN https://developer.mozilla.org/en-US/docs/Web/CSS/gradient/linear-gradient

> By default, if there is no color with a 0% stop, the first color declared will be at that point. Similarly, the last color will continue to the 100% mark, or be at the 100% mark if no length has been declared on that last stop.

may be a bit misleading here. That first sentence might be understood to imply creating an extra stop at 0%, duplicating the first declared color, rather than simply painting that color from its stop position back to 0%; and analogously, although the wording is less ambiguous, an implementer might think they should duplicate the last declared color at 100%. And that would indeed result in the observed behavior. But if my understanding of the actually-expected behavior is correct, I'll suggest rewording this for better clarity.)

Interestingly, in example (d), where the blue stop is moved beyond the right-hand end of the swatch, Chrome and Safari suddenly switch their behavior and now do not use a gradient in the area to the left of the red stop: they turn this into solid color. (Firefox, on the other hand, continues to apply a gradient there, just like in the preceding examples.) I think the Chrome/Safari behavior here is what the spec requires, but I don't understand why they only do this correctly once the blue color stop is beyond the end of the rendered area.

Example (e) shows the equivalent case with the red stop placed beyond the left edge: now, Chrome and Safari render a solid color for the area beyond the blue stop, while Firefox still paints a gradient there.

Examples (f) and (g) serve to illustrate the abrupt change in behavior in Chrome and Safari. The two stops are just shifted right by 1px in (g) compared to (f). In Firefox, this just results in a barely-perceptible shift in the gradient; but in Chrome and Safari it drastically changes the result.

QUESTION 1: Am I right in thinking that any area before the first color stop or after the last color stop should simply render as a solid color, or is longer-hue interpolation expected to somehow "project" into these areas (as the browsers currently do, though not entirely consistently)?

The second group of swatches, (h) through (l), have the two color stops at the exact same position. My reading of the spec is that the correct rendering would have solid red to the left of the stop position, and solid blue to the right. But all three browsers actually render a complete 360° gradient: to the right of the color stops, they render blue-to-blue (all the way around the hue circle), and to the left they render red-to-red (all the way around). For (h) through (j), this result is consistent across all the browsers, and appears to result from adding "implicit stops" at the left and right edges of the swatch, duplicating the color of the nearest real stop.

In example (k), the position of the red and blue stops has been moved beyond the right-hand edge of the swatch. In Firefox, this "stretches" the gradient to the right; it still begins with red at the left edge, but the visible gradient no longer cycles all the way back to red because the stop is outside the swatch area. In Chrome and Safari, on the other hand, the end of the 360° gradient seems to be "clamped" to the right-hand side of the swatch, so (k) looks identical to (j). I can't see any justification for that behavior — though I think the browsers are wrong to be rendering a gradient at all in this case.

Finally, in (l) the stops have been moved before the left-hand edge of the swatch. Again, Firefox stretches the gradient; Chrome "clamps" the left edge, so that (l) is identical to (h). Safari renders this example the same as (j) and (k), which I think is a bug.

QUESTION 2: Is a longer-hue linear gradient with stops at a single position a special case that should render some kind of gradient (as all browsers currently do, regardless of where the stops are positioned), or is that just an implementation bug?

(Note that there are a couple of existing WPT tests that expect this behavior; e.g. see http://wpt.live/css/css-images/gradient/gradient-single-stop-longer-hue-hsl.html. I'm suggesting this may be wrong. Does the web depend on it?)

**Comment by jfkthame** [comment]
To follow up, here is what I think my [codepen example](https://codepen.io/jfkthame/pen/azopZOx) ought to look like, by my reading of the spec:
<img width="198" alt="image" src="https://github.com/user-attachments/assets/e36792a4-60bf-4364-871b-142f25565038" />

Currently, none of the three tested browsers produces this rendering. So if my understanding is correct, and assuming we don't want to modify the spec in the light of what browsers are actually doing, then all three engines have bugs here.

**Comment by schenney-chromium** [comment]
I would be very very worried about changing the appearance of sites that have come to depend on the current behavior, though maybe not many do. Is it possible to craft a query against HTTPArchive that looks for the gradient syntax that you are testing (i.e. gradient styles that hit the conditions of interest).

I would try to change the spec rather than browsers if this turns out to be something that has significant usage. Even though this does look to me like some kind of browser problem with the treatment of identical or near-identical stop colors.

**Comment by jfkthame** [comment]
It's not "identical or near-identical stop colors" that are at issue here; it's what browsers do with `longer hue` gradient interpolation when the first and last defined color stops are not at the 0% and 100% positions. This is an issue whether the gradient is between completely different colors — such as `red` and `blue` in my examples — or from a single color all the way around the hue circle back to the same color, as in the [gradient-single-stop-longer-hue](http://wpt.live/css/css-images/gradient/gradient-single-stop-longer-hue-hsl.html) test.

**Comment by jfkthame** [comment]
> I would try to change the spec rather than browsers

I'd be more ready to be persuaded by this if (a) the existing behavior was fully interoperable between browsers (but it's not, see screenshot in the original issue); and (b) the existing behavior of Blink and Webkit had some clear internal consistency, which I would suggest it lacks.

In what model does it make sense that
```
data:text/html,<div style="width:100px;height:300px;
         background:linear-gradient(in hsl longer hue,red 250px,blue 300px);">
```
and
```
data:text/html,<div style="width:100px;height:300px;
         background:linear-gradient(in hsl longer hue,red 250px,blue 301px);">
```
have the radically different renderings that Chrome and Safari currently give them?
<img width="231" alt="image" src="https://github.com/user-attachments/assets/c0e118fb-6f12-4021-9c41-964c353341a7" />

Surely the only difference between
```
data:text/html,<div style="width:100px;height:300px;
         background:linear-gradient(in hsl shorter hue,red 50px,blue 250px);">
```
and
```
data:text/html,<div style="width:100px;height:300px;
         background:linear-gradient(in hsl longer hue,red 50px,blue 250px);">
```
should be in the region between y=50px and 250px, where the first example takes the "short" arc around the HSL color wheel from red through magenta to blue, and the second example takes the "long" arc via orange, yellow, green, cyan. But there's no logical reason to change what happens outside the bounds of that region, as currently happens:
<img width="230" alt="image" src="https://github.com/user-attachments/assets/5fac63a8-bd5f-47f7-9762-ce4aa5b71ff5" />

**Comment by jfkthame** [comment]
I have gone ahead and filed bugs against all three engines, as ISTM their behavior is a clear violation of the spec, and doesn't really make any sense from an author's point of view either.
- https://bugzilla.mozilla.org/show_bug.cgi?id=1939948
- https://issues.chromium.org/issues/387475844
- https://bugs.webkit.org/show_bug.cgi?id=285382

**Comment by LeaVerou** [comment]
I wonder if [this Chromium bug](https://issues.chromium.org/issues/387013127) that my apprentice @dmitrysharabin recently found could be relevant.

**Comment by jfkthame** [comment]
That looks like an unrelated color-mixing/interpolation bug, I think.

**Comment by yisibl** [comment]
I'm worried that this will break the compatibility of the site.

Currently, there are many places that use `longer hue` to create a rainbow gradient.

```css
.foo {
  background: linear-gradient(to right in hsl longer hue, red 0 0);
}
```

- https://stackoverflow.com/a/75923874/7983605
- https://x.com/jh3yy/status/1879320099858825216
- https://nerdy.dev/easy-hsl-rainbow#:~:text=It%20means%20you%20can%20also%20do%20the%20following
![Image](https://github.com/user-attachments/assets/f937f928-c681-479d-9a5f-a80bf1be94de)


With this proposed change, this fails to produce a gradient and will only show up as a red background.

This breaks compatibility with existing sites and CSS authors must change the CSS above to:

```css
.foo {
  background: linear-gradient(to right in hsl longer hue, red 0 100%);
 /* or */
  background: linear-gradient(to right in hsl longer hue, red, red);
}
```

So, can we relax the restriction so that a gradient can be generated when there is only one color stop.

cc @LeaVerou

**Comment by tabatkins** [comment]
That example isn't one color stop, tho - it's two color stops (two positions with one color, defining two stops).

You say "many sites" are relying on this to generate rainbows, but I don't actually see rainbows very much in practice on the web. Is this just a cute CSS trick that has gotten some shares, or do we have evidence it's *actually* used in a non-trivial number of sites (and, ideally, have evidence that rendering them as a solid color would actually harm the rendering, rather than just *change* the rendering)?

That said, I'm not strongly *opposed* to specifying that we always add one additional color stop at the beginning/end of the list, placed either at 0%/100% or on top of the existing first/last color stop if they're beyond the 0-100% range, and with the same color as the first/last stop. If this does indeed look to be web-required, that's acceptable, but I'd prefer to avoid it if we can, and just fix the browser bugs.

-------

As an aside, we should also probably think about exactly what behavior we *do* want between a double-position stop like that. Double-position stops were added to make it easy to generate solid-color strips in a gradient without having to repeat your color; I suspect that people might expect that to remain true even if they're doing longer-hue interpolation otherwise.

**Comment by jfkthame** [comment]
> As an aside, we should also probably think about exactly what behavior we _do_ want between a double-position stop like that. Double-position stops were added to make it easy to generate solid-color strips in a gradient without having to repeat your color; I suspect that people might expect that to remain true even if they're doing longer-hue interpolation otherwise.

Hmm, yes, I suppose that might take people by surprise.

The [spec text](https://drafts.csswg.org/css-images-4/#color-stop-syntax) currently says that:

> A [color stop](https://drafts.csswg.org/css-images-4/#color-stop) with two positions is equivalent to specifying two color stops with the same color, one for each position.

which means that `linear-gradient(in hsl longer hue, red 0% 100%)` should be exactly equivalent to `linear-gradient(in hsl longer hue, red 0%, red 100%)` (or indeed to `linear-gradient(in hsl longer hue, red, red)`). I'd be a bit reluctant to break that equivalence by giving a "double-position stop" some kind of special handling, distinct from two separately-defined stops with the same color.

The spec goes on to mention that:

> Specifying two locations makes it easier to create solid-color "stripes" in a gradient, without having to repeat the color twice.

Perhaps a note should be added here clarifying that _this does not apply to `longer hue` gradients_, because it is inherent in longer-hue gradients that between one stop and a following stop of the same color, there will be a complete "cycle" of color rather than a band of solid color.

**Comment by tabatkins** [comment]
> I'd be a bit reluctant to break that equivalence by giving a "double-position stop" some kind of special handling, distinct from two separately-defined stops with the same color.

Yeah, on further thought I think I'd prefer to avoid complicating this, too. The note just needs an amendment, as you say.

**Comment by yisibl** [comment]
> Perhaps a note should be added here clarifying that this does not apply to longer hue gradients, because it is inherent in longer-hue gradients that between one stop and a following stop of the same color, there will be a complete "cycle" of color rather than a band of solid color.

I'm definitely in favor of it! Also, hopefully some examples can be added for rainbow gradients to avoid more misunderstandings.

**Comment by yisibl** [comment]
After rethinking, I also agree that it would be better for browsers to fix it, which prevents the specification from getting complicated and reduces the learning cost for users.

Looking back, since all three browsers implemented the specification incorrectly. This led to a misunderstanding by many CSS authors that `in hsl longer hue, red 0 0` could produce gradients.

Thankfully, this feature may not be used by many people at the moment, and it's not too late.

@CGQAQ and I will be working on fixing this in Chrome(See [this patch](https://chromium-review.googlesource.com/c/chromium/src/+/6253537)). And possibly add a **UseCounters** to count exactly how many sites are affected. If necessary, we can issue a warning in the console ⚠️.

In addition, I've collected some data to make it easier to assess the impact.

## Potentially Affected Sites

Here are some real use cases found through GitHub searches:

| Site | CSS | Source Code Link | Screenshot |
| --- | --- | --- | --- |
| [https://dreameduinfo.com/](https://dreameduinfo.com/)<br/>@raselmahmud-coder | `conic-gradient(   from var(--bg-angle) in oklch longer hue,    oklch(0.85 0.37 0) 0 0    )` | [open](https://github.com/raselmahmud-coder/FrontEndDreamEdu/blob/175d068a8c8d8cbaa60e7d18d21177d7bc9ac3c9/src/styles/AnimationCustom.module.css#L42-L45) | ![Image](https://github.com/user-attachments/assets/f5840cb6-e9e8-4d65-8c5e-3835e7d438c9) |
| [https://leanrada.com/htmz](https://leanrada.com/htmz)<br/>@Kalabasa | `linear-gradient(135deg in hsl longer hue, red, red 50%)` | [open](https://github.com/gmh5225/htmz/blob/2c9665db9cd49fe03717f8b21ab884c27c942144/docs/style.css#L319-L326) | ![Image](https://github.com/user-attachments/assets/445be5ea-eb58-473c-9f77-c79948bae3c7) |


## Affected front-end build tools

Starting with [esbuild 0.19.9](https://github.com/evanw/esbuild/blob/main/CHANGELOG-2023.md#0199), support for converting gradients with `in hsl longer hue` to equivalent normal gradients.

```css
/* input */
.test-longer {
  background: linear-gradient(in hsl longer hue, red);
}

/* esbuild output */
.test-longer {
  background: linear-gradient(red, #ffbf00 12.5%, #ffef00, #dfff00 18.75%, #7fff00, #20ff00 31.25%, #00ff10, #00ff40 37.5%, #0ff, #0040ff 62.5%, #0010ff, #2000ff 68.75%, #7f00ff, #df00ff 81.25%, #ff00ef, #ff00bf 87.5%, red);
}
```

Opening [esbuild's test page](https://esbuild.github.io/gradient-tests/) in Firefox 136+, we found that this compatibility is broken.

I can file an issue for `esbuild` at a later date and I believe it will be fixed soon.

![Image](https://github.com/user-attachments/assets/90632b72-ba88-4873-9e09-e1477ea8f37b)

`@csstools/postcss-gradients-interpolation-method` may also be affected, I haven't tested it yet.

**Comment by yisibl** [comment]
@jfkthame For cases smaller than 1px, Firefox's rendering is not quite the same.

![Image](https://github.com/user-attachments/assets/a7cfe71f-d69d-4853-9beb-1b9ebc44e38b)

```html
data:text/html;charset=UTF-8,<!DOCTYPE html>
<style> div {
  width: 200px; height: 50px; border: 1px solid black;
  background:linear-gradient(90deg in hsl longer hue, 
  	transparent 0 var(--x), red var(--x) 100px, transparent 0 100%);
  --x: 99px;
}
</style>
<div></div>
```

**Comment by tabatkins** [comment]
All right, at least for now I've updated the note about double-position creating a stripe, and added a note emphasizing that the spec means what it says about the color before the first stop and after the last stop. ^_^

Agenda+ to confirm that this should be closed No Change otherwise.

**Comment by svgeesus** [comment]
[That change](https://github.com/w3c/csswg-drafts/commit/43c8b654b8a93319bb8b3b059613cc616878c9a3) looks correct to me.

**Comment by johannesmutter** [comment]
_"in hsl longer hue"_ worked before in Chrome Canary, but now not anymore, just renders red. Safari still works fine.
```css
conic-gradient(in hsl longer hue, red 0 0);
```
I used this in a color picker, [simplified version here](https://svelte.dev/playground/72354dbd43f14f4ba4e92ea536975d30?version=5.28.2#H4sIAAAAAAAAE61WW6_aRhD-KyMnKNBgMKTkpC4cKY2OlD5UqvrSVqUii3ewt2dZW7tjLkL892p842AuR5X6YHs1883Mzt0Hz4g1eqH3JdWphd8TRO28vrdSGp0X_nXwaJ8xnwlev0Z_zrKB26Ampi2Fw2v0KDWEhpwXelMXWZXR49wAaCRIcoQZvHUkCLtB78cX9MWW77CwuDohTgDlFus0d7iQ6dacACuhHTKKcavcRKRSA3kmBeHXHLu4QUM9ODAbIEqNI7AYEczOTQ5ipJ_S3Ehl4i9aoaHfMKLKfi0YoSG0ix3MCh0DjSuC9-V5qyQlMITxNYl9LUFpVgskqOKELiQkay9uPYiKe_wBfmP4HLlvIf88IfcVkiMnTKw56L8ISgaChBl35b4PcteD72D0KYBhyfr150qofA-H8Fn-k7taAaXgSFiClU3XQAlTMhBGQpxCpNPoeasclrK1yW558GH8EMB7-PAx6EGHP2eWypoooAX92MpmIozU-CL9ray2S4NsjpWBdiG8rn-dbtpVo1bQPbPRcG4YKEy8YifPujcdKIr69atqFBv8D1q4GJyg3ArWtYitkFw23EwSrdqg7H5rCkxFfg3o1s4mTnffHpIcj30YBUGnD5Og04NAYty_hnmY1JAfbmImjZrRp5ugcaNo_HAT1Cj68JExJaTHn2_1gOAI6HytjDAR_i8BuOt_CRlP7gbgRSDvRKBEPUzuh6BE8ftWFKbDZiDPzXQtlClG83SZE6WmasilMjKkRLnZ4WxIHqvutkr4WixRz95FvD38AvGuko60cG42916wuMXnXsVPTVGmXKWzw0VvH89AedaC5Nk5gNu1BWHSOaholBaqoBWwx-mw9L6MhFSbqy6cOmfugaO9xnApoufY8taYHa401vFxOpRqc19tU45XtV4W60npdFilb26mhSQfqUrkgc9zElqHkBuHxB1Ahb8sVc-MtbCxMiHUA3mtjF8uppDLaJNUdKlcpsU-hJXGegsJrWLjK8K1C6u9U3F4a6jV3q_-A1rck38hvHl6eqrIschCGAdZrZ4t-VvLVH7Xgwxg8DJ8lR_F9g1hNDnJC5dhRH6RkxBGtfHUSrQ-BzN3Ibddw9j5LhEy3YYwynbF8322gzdBEExK4-W-GvC6Oly60hoZynBngk5NjJbrvw8WJQRQ_PRwbqpuya1LbQhZqqoYMa_MjXsOgfUJfTmKImUjjSDqf4xmHpAVxmXC8mALrlIjoaPuJOiAD5Ns12swSy2i51e4k6BTjZQ6IdNhWXxe3yPckRfy-j3-3fdIKL1VRnphsYmO_wLbknGzcgoAAA==)

| Chrome | Safari |
| -------- | ------- |
| ![Image](https://github.com/user-attachments/assets/5e5c76d6-4c2f-41e2-bc68-df39d8828e6e) |  ![Image](https://github.com/user-attachments/assets/204a0f25-37a3-4b75-9dcd-718f1842b522) |

**Comment by tabatkins** [comment]
Yes, the new Chrome behavior is intentional, and is exactly what this issue was about. Use `conic-gradient(in hsl longer hue, red 0 100%)` instead.

**Comment by smfr** [comment]
I'm quite surprised that `background: linear-gradient(to right in hsl longer hue, green 25%, red 25%, red 75%, pink 75%);` cycles through all the colors between those two "red" stops. How are authors supposed to write "longer hue" gradients with sections of solid color?

**Comment by facelessuser** [comment]
> How are authors supposed to write "longer hue" gradients with sections of solid color?

Unfortunately, there isn't really a way using "longer" as it always force the interpolation path to take the longer arc.

I personally always thought this to be a great case for undefined hues, but that only works if you resolve the undefined hue post-interpolation. CSS chose to resolve the hue pre-interpolation. Resolving before interpolation creates two defined hues with a long arc traveling 360 degrees. I always thought it odd to treat an interpolation of an undefined and defined hue as having an arc, but I lost that argument 🙂.

I personally chose to treat the CSS approach as its own thing so I could utilize undefined hues in a non-CSS approach that seemed more intuitive (at least to me).

![Image](https://github.com/user-attachments/assets/18a89fac-233c-4c96-8d5e-0f72fe092d94)

I think inteprolation between two identical defined hues creating a rainbow should be expected. It's actually an easy way to create rainbow gradients.

**Comment by yisibl** [comment]
> I'm quite surprised that `background: linear-gradient(to right in hsl longer hue, green 25%, red 25%, red 75%, pink 75%);` cycles through all the colors between those two "red" stops. How are authors supposed to write "longer hue" gradients with sections of solid color?

For now, the only option is to remove the `longer hue` or replace it with `shorter hue`.

**Comment by svgeesus** [comment]
> Agenda+ to confirm that this should be closed No Change otherwise.

@tabatkins It seems there is consensus, does this still need Agenda+ or can we close?

**Comment by tabatkins** [comment]
> How are authors supposed to write "longer hue" gradients with sections of solid color?

As @facelessuser says, there is currently no way to do so. What we might want to do is allow hue interpolation to be specified between stops (alongside the interpolation hint), like we're now doing with the design of the `*-interpolate()` functions. Like those functions, the interpolation method specified in the function's prelude would give the default, but it would be overridable between any particular pair.

That'd be a separate issue, tho.

> It seems there is consensus, does this still need Agenda+ or can we close?

Simon's comment from a few years ago suggests there's not necessarily obvious consensus, so I'd still like to get this approved by the WG.

**Comment by smfr** [comment]
Not sure if you're referring to one my comments. But I'm fixing the webkit bug so I think we don't need to discuss again.

**Comment by tabatkins** [comment]
lolllll "years ago" I meant "hours ago". Sorry about that.

Anyway if you're cool with closing No Change, that works for me.
